persistence/src/main/java/edu/unc/lib/dl/util/MultiDigestInputStreamWrapper.java

```diff
@@ -0,0 +1,156 @@
+        if (wantDigests != null) {
+            for (DigestAlgorithm wantDigest : wantDigests) {
+                if (!algToDigest.containsKey(wantDigest)) {
+                    algToDigest.put(wantDigest, null);
+                }
+            }
+        }
+    }
+
+    /**
+     * Get the InputStream wrapped to produce the requested digests
+     *
+     * @return wrapped input stream
+     */
+    public InputStream getInputStream() {
+        streamRetrieved = true;
+        InputStream digestStream = sourceStream;
+        for (DigestAlgorithm algorithm : algToDigest.keySet()) {
+            try {
+                // Progressively wrap the original stream in layers of digest streams
+                digestStream = new DigestInputStream(
+                        digestStream, MessageDigest.getInstance(algorithm.getName()));
+            } catch (final NoSuchAlgorithmException e) {
+                throw new UnsupportedAlgorithmException("Unsupported digest algorithm: " + algorithm, e);
+            }
+
+            algToDigestStream.put(algorithm, (DigestInputStream) digestStream);
+        }
+        return digestStream;
+    }
+
+    /**
+     * After consuming the inputstream, verify that all of the computed digests
+     * matched the provided digests.
+     *
+     * Note: the wrapped InputStream will be consumed if it has not already been read.
+     *
+     * @throws InvalidChecksumException thrown if any of the digests did not match
+     */
+    public void checkFixity() throws InvalidChecksumException {
+        calculateDigests();
+
+        algToDigest.forEach((algorithm, originalDigest) -> {
+            // Skip any algorithms which were calculated but no digest was provided for verification
+            if (originalDigest == null) {
+                return;
+            }
+            final String computed = computedDigests.get(algorithm);
+
+            if (!originalDigest.equalsIgnoreCase(computed)) {
+                throw new InvalidChecksumException(String.format(
+                        "Checksum mismatch, computed %s digest %s did not match expected value %s",
+                        algorithm, computed, originalDigest));
+            }
+        });
+
+    }
+
+    /**
+     * Returns the list of digests calculated for the wrapped InputStream
+     *
+     * Note: the wrapped InputStream will be consumed if it has not already been read.
+     *
+     * @return list of digests calculated from the wrapped InputStream, in URN format.
+     */
+    public Map<DigestAlgorithm, String> getDigests() {
+        calculateDigests();
+
+        return computedDigests;
+    }
+
+    private void calculateDigests() {
+        if (computedDigests != null) {
+            return;
+        }
+
+        if (!streamRetrieved) {
+            // Stream not previously consumed, consume it now in order to calculate digests
+            try (final InputStream is = getInputStream()) {
```

Comment:
What is the purpose of reading through the file here? I don't see "is used anywhere else.

Reply:
Its because the input stream is wrapped in a DigestInputStream
https://github.com/UNC-Libraries/Carolina-Digital-Repository/pull/1022/files#diff-7e1c346f26b46834efa5d9ed91646b62R83-R90
When you read the contents of the stream, it calculates one or more digests. You could consume the contents of the stream for some other purpose, like writing the stream to a file or something, and it would still calculate the digests. That is why this block is wrapped in `if (!streamRetrieved)`, so that it won't try to re-read the stream. But if we had not already read the stream at the time that digests are needed, then we have to read it. This is the case for the FixityCheckJob since it is loading the files just for the sake of checksumming.